// @mallikgalibshahriar/failsafe-report/dist/reporters/notifications.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NotificationService = void 0;
const axios_1 = __importDefault(require("axios"));
class NotificationService {
    static async send(report, config) {
        switch (config.type) {
            case 'slack':
                await this.sendToSlack(report, config.webhookUrl);
                break;
            case 'teams':
                await this.sendToTeams(report, config.webhookUrl);
                break;
            default:
                throw new Error(`Unsupported notification type: ${config.type}`);
        }
    }
    static async sendToSlack(report, webhookUrl) {
        const summary = report.summary;
        const statusEmoji = summary.failed > 0 ? '❌' : '✅';
        const blocks = [
            {
                type: 'header',
                text: {
                    type: 'plain_text',
                    text: `${statusEmoji} Test Report: ${summary.framework.toUpperCase()}`,
                    emoji: true
                }
            },
            {
                type: 'section',
                fields: [
                    { type: 'mrkdwn', text: `*Total:* ${summary.total}` },
                    { type: 'mrkdwn', text: `*Passed:* ${summary.passed} ✅` },
                    { type: 'mrkdwn', text: `*Failed:* ${summary.failed} ❌` },
                    { type: 'mrkdwn', text: `*Duration:* ${(summary.duration / 1000).toFixed(2)}s` }
                ]
            }
        ];
        if (report.insights && report.insights.length > 0) {
            blocks.push({
                type: 'section',
                text: {
                    type: 'mrkdwn',
                    text: `*Actionable Insights:*\n• ${report.insights.join('\n• ')}`
                }
            });
        }
        if (report.clusters && report.clusters.length > 0) {
            const topFailures = report.clusters.slice(0, 3).map(c => `*${c.count}x* \`${c.message}\``).join('\n');
            blocks.push({
                type: 'section',
                text: {
                    type: 'mrkdwn',
                    text: `*Top Failures:*\n${topFailures}`
                }
            });
        }
        try {
            await axios_1.default.post(webhookUrl, { blocks });
        }
        catch (err) {
            console.error('Failed to send Slack notification:', err);
        }
    }
    static async sendToTeams(report, webhookUrl) {
        const summary = report.summary;
        const card = {
            type: 'message',
            attachments: [{
                    contentType: 'application/vnd.microsoft.card.adaptive',
                    content: {
                        type: 'AdaptiveCard',
                        body: [
                            { type: 'TextBlock', size: 'Medium', weight: 'Bolder', text: `Test Report: ${summary.framework}` },
                            {
                                type: 'FactSet', facts: [
                                    { title: 'Total', value: `${summary.total}` },
                                    { title: 'Passed', value: `${summary.passed}` },
                                    { title: 'Failed', value: `${summary.failed}` }
                                ]
                            }
                        ],
                        $schema: 'http://adaptivecards.io/schemas/adaptive-card.json',
                        version: '1.2'
                    }
                }]
        };
        try {
            await axios_1.default.post(webhookUrl, card);
        }
        catch (err) {
            console.error('Failed to send Teams notification:', err);
        }
    }
}
exports.NotificationService = NotificationService;
