// @mallikgalibshahriar/failsafe-report/dist/parsers/parsers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MochaParser = exports.VitestParser = exports.CypressParser = exports.PlaywrightParser = exports.JestParser = exports.BaseParser = void 0;
class BaseParser {
}
exports.BaseParser = BaseParser;
class JestParser extends BaseParser {
    constructor() {
        super(...arguments);
        this.name = 'jest';
    }
    canParse(content) {
        return content && content.numTotalTests !== undefined && content.testResults !== undefined;
    }
    parse(content) {
        const results = [];
        content.testResults.forEach((suite) => {
            suite.assertionResults.forEach((test) => {
                results.push({
                    id: `${suite.name}-${test.title}`,
                    name: test.title,
                    suite: suite.name,
                    status: test.status === 'passed' ? 'passed' : test.status === 'failed' ? 'failed' : 'skipped',
                    duration: test.duration || 0,
                    error: test.failureMessages && test.failureMessages.length > 0 ? {
                        message: test.failureMessages[0],
                        stack: test.failureMessages.join('\n')
                    } : undefined
                });
            });
        });
        const summary = {
            total: content.numTotalTests,
            passed: content.numPassedTests,
            failed: content.numFailedTests,
            skipped: content.numPendingTests,
            flaky: 0, // Jest doesn't natively report flakiness in the basic JSON output without extra tags
            duration: Date.now() - content.startTime, // approximate or from content if available
            framework: 'jest',
            timestamp: new Date().toISOString()
        };
        return { summary, results };
    }
}
exports.JestParser = JestParser;
class PlaywrightParser extends BaseParser {
    constructor() {
        super(...arguments);
        this.name = 'playwright';
    }
    canParse(content) {
        return content && content.config && content.suites;
    }
    parse(content) {
        const results = [];
        const walkSuites = (suites) => {
            suites.forEach(suite => {
                if (suite.specs) {
                    suite.specs.forEach((spec) => {
                        spec.tests.forEach((test) => {
                            const result = test.results[0];
                            results.push({
                                id: spec.id || `${suite.title}-${spec.title}`,
                                name: spec.title,
                                suite: suite.title,
                                status: result.status === 'expected' ? 'passed' : result.status === 'unexpected' ? 'failed' : 'skipped',
                                duration: result.duration || 0,
                                error: result.error ? {
                                    message: result.error.message,
                                    stack: result.error.stack
                                } : undefined
                            });
                        });
                    });
                }
                if (suite.suites)
                    walkSuites(suite.suites);
            });
        };
        walkSuites(content.suites);
        const summary = {
            total: results.length,
            passed: results.filter(r => r.status === 'passed').length,
            failed: results.filter(r => r.status === 'failed').length,
            skipped: results.filter(r => r.status === 'skipped').length,
            flaky: 0,
            duration: content.stats?.duration || 0,
            framework: 'playwright',
            timestamp: new Date().toISOString()
        };
        return { summary, results };
    }
}
exports.PlaywrightParser = PlaywrightParser;
class CypressParser extends BaseParser {
    constructor() {
        super(...arguments);
        this.name = 'cypress';
    }
    canParse(content) {
        return content && content.results && content.stats && content.stats.suites !== undefined;
    }
    parse(content) {
        const results = [];
        content.results.forEach((suite) => {
            suite.tests.forEach((test) => {
                results.push({
                    id: test.uuid || `${suite.title}-${test.title}`,
                    name: test.title,
                    suite: suite.title,
                    status: test.state === 'passed' ? 'passed' : test.state === 'failed' ? 'failed' : 'skipped',
                    duration: test.duration || 0,
                    error: test.err && Object.keys(test.err).length > 0 ? {
                        message: test.err.message,
                        stack: test.err.stack
                    } : undefined
                });
            });
        });
        const summary = {
            total: content.stats.tests,
            passed: content.stats.passes,
            failed: content.stats.failures,
            skipped: content.stats.pending,
            flaky: 0,
            duration: content.stats.duration,
            framework: 'cypress',
            timestamp: content.stats.start || new Date().toISOString()
        };
        return { summary, results };
    }
}
exports.CypressParser = CypressParser;
class VitestParser extends BaseParser {
    constructor() {
        super(...arguments);
        this.name = 'vitest';
    }
    canParse(content) {
        return content && content.numTotalTests !== undefined && content.testResults !== undefined && content.startTime !== undefined;
    }
    parse(content) {
        const results = [];
        content.testResults.forEach((suite) => {
            suite.assertionResults.forEach((test) => {
                results.push({
                    id: `${suite.name}-${test.title}`,
                    name: test.title,
                    suite: suite.name,
                    status: test.status === 'passed' ? 'passed' : test.status === 'failed' ? 'failed' : 'skipped',
                    duration: test.duration || 0,
                    error: test.failureMessages && test.failureMessages.length > 0 ? {
                        message: test.failureMessages[0],
                        stack: test.failureMessages.join('\n')
                    } : undefined
                });
            });
        });
        const summary = {
            total: content.numTotalTests,
            passed: content.numPassedTests,
            failed: content.numFailedTests,
            skipped: content.numPendingTests,
            flaky: 0,
            duration: Date.now() - content.startTime,
            framework: 'vitest',
            timestamp: new Date().toISOString()
        };
        return { summary, results };
    }
}
exports.VitestParser = VitestParser;
class MochaParser extends BaseParser {
    constructor() {
        super(...arguments);
        this.name = 'mocha';
    }
    canParse(content) {
        return content && content.stats && content.tests && Array.isArray(content.tests);
    }
    parse(content) {
        const results = content.tests.map((test) => ({
            id: test.fullTitle,
            name: test.title,
            suite: test.fullTitle.replace(test.title, '').trim(),
            status: test.err && Object.keys(test.err).length > 0 ? 'failed' : 'passed',
            duration: test.duration || 0,
            error: test.err && Object.keys(test.err).length > 0 ? {
                message: test.err.message,
                stack: test.err.stack
            } : undefined
        }));
        const summary = {
            total: content.stats.tests,
            passed: content.stats.passes,
            failed: content.stats.failures,
            skipped: content.stats.pending,
            flaky: 0,
            duration: content.stats.duration,
            framework: 'mocha',
            timestamp: content.stats.start || new Date().toISOString()
        };
        return { summary, results };
    }
}
exports.MochaParser = MochaParser;
